src/mappings/entities/mapConfidentialAsset.ts

```diff
@@ -0,0 +1,217 @@
+    updatedBlockId: blockId,
+  }).save();
+};
+
+const handleVenuesAllowed = async (blockId: string, params: Codec[]): Promise<void> => {
+  const [, rawAssetId, rawVenueId] = params;
+
+  const assetId = getTextValue(rawAssetId);
+  const venuesAllowed = JSON.parse(rawVenueId.toString()).map(val => getNumberValue(val));
+
+  const asset = await ConfidentialAsset.get(assetId);
+
+  const foundAllowedVenues: ConfidentialAssetAllowedVenue[] = (
+    await Promise.all(
+      venuesAllowed.map(venueId => ConfidentialAssetAllowedVenue.get(`${assetId}/${venueId}`))
+    )
+  ).filter(venue => venue !== undefined);
+
+  const foundAllowedVenuesUpdatePromises = foundAllowedVenues
+    .filter(venue => !venue.isAllowed)
+    .map(venue => {
+      venue.isAllowed = true;
+      venue.updatedBlockId = blockId;
+
+      return venue.save();
+    });
+
+  const foundVenuesIds = foundAllowedVenues.map(venue => venue.venueId);
+
+  const venuesRelationsToCreate = venuesAllowed.filter(
+    venueId => !foundVenuesIds.includes(venueId)
+  );
+
+  const allowedVenuesPromises = venuesRelationsToCreate.map(venueId =>
+    ConfidentialAssetAllowedVenue.create({
+      id: `${assetId}/${venueId}`,
+      venueId,
+      assetId,
+      isAllowed: true,
+      createdBlockId: blockId,
+      updatedBlockId: blockId,
+    }).save()
+  );
+
+  await Promise.all([...foundAllowedVenuesUpdatePromises, ...allowedVenuesPromises]);
+
+  if (asset) {
+    asset.updatedBlockId = blockId;
+
+    await asset.save();
+  }
+};
+
+const handleVenuesBlocked = async (blockId: string, params: Codec[]): Promise<void> => {
+  const [, rawAssetId, rawVenueId] = params;
+
+  const assetId = getTextValue(rawAssetId);
+  const blockedVenues = JSON.parse(rawVenueId.toString()).map(val => getNumberValue(val));
+
+  const foundAllowedVenues: ConfidentialAssetAllowedVenue[] = (
+    await Promise.all(
+      blockedVenues.map(venueId => ConfidentialAssetAllowedVenue.get(`${assetId}/${venueId}`))
+    )
+  ).filter(venue => venue !== undefined);
+
+  const foundAllowedVenuesUpdatePromises = foundAllowedVenues
+    .filter(venue => venue.isAllowed)
+    .map(venue => {
+      venue.isAllowed = false;
+      venue.updatedBlockId = blockId;
+
+      return venue.save();
+    });
+
+  const foundVenuesIds = foundAllowedVenues.map(venue => venue.venueId);
+
+  const venuesRelationsToCreate = blockedVenues.filter(
+    venueId => !foundVenuesIds.includes(venueId)
+  );
+
+  const blockedVenuesPromises = venuesRelationsToCreate.map(venueId =>
+    ConfidentialAssetAllowedVenue.create({
+      id: `${assetId}/${venueId}`,
+      venueId,
+      assetId,
+      isAllowed: false,
+      createdBlockId: blockId,
+      updatedBlockId: blockId,
+    }).save()
+  );
+
+  await Promise.all([...foundAllowedVenuesUpdatePromises, ...blockedVenuesPromises]);
+
+  const asset = await ConfidentialAsset.get(assetId);
+
+  if (asset) {
+    asset.updatedBlockId = blockId;
+
+    await asset.save();
+  }
+};
```

Comment:
The logic can be simplified here.

Note - For venues allowed/blocked, the venue doesn't have to exists (For example you can allow say venue id 9999 even if doesn't exists. So we can only track the IDs here. That is why I recommended to remove the entity and only add a single attribute as `allowedVenues` to `ConfidentialAsset`. 

Also, if we block a venue say 45 and it is not already allowed, nothing happens (list remains unaltered). In short, we only keep a track of allowed venues and if the blocked venues are a part of allowed we remove that from the list, otherwise do nothing